File: V.2/Player.cpp
#include "Player.h"
#include "Entity.h"
#include "Location.h"
#include <iostream>

Player::Player(const Location& loc, const Vector& vec, const char& model, const char& tail, int tailsize)
	: Entity(loc, vec, model), tail(tail), tailsize(tailsize)
{
	
}

void Player::tick()
{
	Location oldloc = this->loc;
	Entity::tick();

	if (oldloc != this->loc)
	{
		this->addTail(oldloc);
	}
}

void Player::addTail(const Location& tloc)
{
	if (this->tailLocs.size() == this->tailsize)
	{
		this->tailLocs.erase(this->tailLocs.begin());
	}
	
	this->tailLocs.insert(this->tailLocs.end(), *(new Location(tloc.x, tloc.y)));
}

void Player::render(int x, int y)
{
	Entity::render(x, y);

	for (Location& loc : this->tailLocs)
	{
		if (loc.x == x && loc.y == y && loc != this->loc)
		{
			std::cout << this->tail;
		}
	}
}

Player::~Player()
{
	for (Location loc : this->tailLocs)
	{
		delete &loc;
	}
}

File: SnakeCS/Player.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections;

namespace SnakeCS
{
    class Player : Entity
    {
        char Tail;
        int TailSize;

        public ArrayList tailLocs = new ArrayList();

        public Player(Location location, Vector vector, char model, char Tail, int TailSize) : base(location, vector, model)
        {
            this.Tail = Tail;
            this.TailSize = TailSize;
        }

        public override void Tick()
        {
            Location oldloc = new Location(base.location.x, base.location.y);
            base.Tick();
            this.AddTail(oldloc);
        }

        public void AddTail(Location location)
        {
            if (tailLocs.Count >= this.TailSize)
            {
                tailLocs.RemoveAt(0);
            }
            tailLocs.Add(location);
        }

        public override void Render(int x, int y)
        {
            base.Render(x, y);
            foreach (Location loc in tailLocs)
            {
                if (loc.x == x && loc.y == y)
                {
                    Console.Write(this.Tail);
                }
            }
        }
    }
}
